network/router.go

```diff
@@ -0,0 +1,394 @@
+	if err != nil {
+		log.Lvl2(r.address, "Couldn't send to", e, ":", err, "trying again")
+		c, err := r.connect(e)
+		if err != nil {
+			return err
+		}
+		err = c.Send(msg)
+		if err != nil {
+			return err
+		}
+	}
+	log.Lvl5("Message sent")
+	return nil
+}
+
+func (r *Router) connect(si *ServerIdentity) (Conn, error) {
+	c, err := r.host.Connect(si.Address)
+	if err != nil {
+		return nil, err
+	}
+	if err := r.negotiateOpen(si, c); err != nil {
+		return nil, err
+	}
+
+	r.launchHandleRoutine(si, c)
+	return c, nil
+
+}
+
+// handleConn is the main routine for a connection to wait for incoming
+// messages.
+func (r *Router) handleConn(remote *ServerIdentity, c Conn) {
+	defer func() {
+		// when leaving, unregister the connection
+		if err := c.Close(); err != nil {
+			log.Error(r.address, "having error closing conn to ", remote.Address, ":", err)
+		}
+		// and release one on the waitgroup
+		r.wg.Done()
+	}()
+	address := c.Remote()
+	log.Lvl3(r.address, "Handling new connection to ", remote.Address)
+	for {
+		packet, err := c.Receive()
+		// So the receiver can know about the error
+		packet.SetError(err)
+		packet.From = address
+		packet.ServerIdentity = remote
+
+		// whether the router is closed
+		if r.Closed() {
+			// signal we are done with this go routine.
+			return
+		}
+
+		if err != nil {
+			// something went wrong on this connection
+			log.Lvlf4("%+v got error (%+s) while receiving message", r.id.String(), err)
+
+			if err == ErrClosed || err == ErrEOF || err == ErrTemp {
+				// remote connection closed
+				log.Lvl3(r.address, "handleConn with closed connection: stop (dst=", remote.Address, ")")
+				return
+			}
+			// weird error let's try again
+			log.Lvl3(r.id, "Error with connection", address, "=>", err)
+			continue
+		}
+
+		if err := r.Dispatch(&packet); err != nil {
+			log.Lvl3("Error dispatching:", err)
+		}
+	}
+}
+
+// connection returns the connection associated with this ServerIdentity. Nil if
+// nothing found. It always return the first connection associated.
+func (r *Router) connection(sid ServerIdentityID) Conn {
+	r.connsMut.Lock()
+	defer r.connsMut.Unlock()
+	arr := r.connections[sid]
+	if len(arr) == 0 {
+		return nil
+	}
+	return arr[0]
+}
+
+// registerConnection registers an ServerIdentity for a new connection, mapped with the
+// real physical address of the connection and the connection itself
+// it locks (and unlocks when done):  networkLock
+func (r *Router) registerConnection(remote *ServerIdentity, c Conn) {
+	log.Lvl4(r.address, "Registers", remote.Address)
+	r.connsMut.Lock()
+	defer r.connsMut.Unlock()
+	_, okc := r.connections[remote.ID]
+	if okc {
+		log.Lvl5("Connection already registered. Appending new connection to same identity.")
+	}
+	r.connections[remote.ID] = append(r.connections[remote.ID], c)
+}
+
+func (r *Router) reset() {
+	r.closedMut.Lock()
+	r.isClosed = false
+	r.closedMut.Unlock()
+}
+
+func (r *Router) launchHandleRoutine(dst *ServerIdentity, c Conn) {
```

Comment:
I had quite a hard time figuring out how to make sure every go routines finishes when calling `Stop()`: I use a `WaitGroup`, and the `wg.Add(1)` *must* be done before launching the go routine (see [godoc](https://golang.org/pkg/sync/#WaitGroup). And this is launched at two places so .. :)